src/client/datascience/jupyter/kernels/helpers.ts

```diff
@@ -1774,3 +1793,98 @@ export async function executeSilently(session: IJupyterSession, code: string): P
 
     return outputs;
 }
+async function switchController(
+    resource: Resource,
+    serviceContainer: IServiceContainer
+): Promise<VSCodeNotebookController | undefined> {
+    const commandManager = serviceContainer.get<ICommandManager>(ICommandManager);
+    const notebooks = serviceContainer.get<IVSCodeNotebook>(IVSCodeNotebook);
+    const editor = findNotebookEditor(resource, notebooks, serviceContainer.get(IInteractiveWindowProvider));
+
+    // Listen for selection change events (may not fire if user cancels)
+    const controllerManager = serviceContainer.get<INotebookControllerManager>(INotebookControllerManager);
+    let controller: VSCodeNotebookController | undefined;
+    const waitForSelection = createDeferred<VSCodeNotebookController>();
+    const disposable = controllerManager.onNotebookControllerSelected((e) => waitForSelection.resolve(e.controller));
+
+    const selected = (await selectKernel(
+        resource,
+        notebooks,
+        serviceContainer.get(IInteractiveWindowProvider),
+        commandManager
+    )) as boolean;
+    if (selected && editor) {
+        controller = await waitForSelection.promise;
+    }
+    disposable.dispose();
+    return controller;
+}
+
+export async function connectToKernel(
+    initialController: VSCodeNotebookController,
+    serviceContainer: IServiceContainer,
+    resource: Resource,
+    notebook: NotebookDocument
+): Promise<IKernel> {
+    const kernelProvider = serviceContainer.get<IKernelProvider>(IKernelProvider);
+    const errorHandler = serviceContainer.get<IDataScienceErrorHandler>(IDataScienceErrorHandler);
+    const memento = serviceContainer.get<Memento>(IMemento, GLOBAL_MEMENTO);
+    let kernel: IKernel | undefined;
+    let controller: VSCodeNotebookController | undefined = initialController;
+    while (kernel === undefined) {
+        // Try to create the kernel (possibly again)
+        kernel = kernelProvider.getOrCreate(notebook, {
+            metadata: controller.connection,
+            controller: controller.controller,
+            resourceUri: resource
+        });
+
+        try {
+            await kernel.start();
+        } catch (error) {
+            if (controller.connection.interpreter) {
+                // If we failed to start the kernel, then clear cache used to track
+                // whether we have dependencies installed or not.
+                // Possible something is missing.
+                clearInstalledIntoInterpreterMemento(
+                    memento,
+                    undefined,
+                    controller.connection.interpreter.path
+                ).ignoreErrors();
+            }
+
+            const handleResult = await errorHandler.handleKernelError(error, 'start', controller.connection, resource);
+            switch (handleResult.kind) {
+                case 'Canceled':
+                    throw new CancellationError(
```

Comment:
Don't we need to dispose the kernle here as well?
Basically the kernel hasn't been started

Reply:
Yes we do. Good catch